daemon/daemon.go

```diff
@@ -681,49 +687,135 @@ func (d *Daemon) Stop(sigCh chan<- os.Signal) error {
 	}
 
 	if restartSystem {
-		// ask for shutdown and wait for it to happen.
-		// if we exit snapd will be restared by systemd
-		rebootDelay := 1 * time.Minute
+		return d.doReboot(sigCh, rebootWaitTimeout)
+	}
+
+	if d.restartSocket {
+		return ErrRestartSocket
+	}
+
+	return nil
+}
+
+func (d *Daemon) rebootDelay() (time.Duration, error) {
+	d.state.Lock()
+	defer d.state.Unlock()
+	now := time.Now()
+	// see if had already scheduled one
+	var rebootAt time.Time
+	err := d.state.Get("daemon-system-restart-at", &rebootAt)
+	if err != nil && err != state.ErrNoState {
+		return 0, err
+	}
+	rebootDelay := 1 * time.Minute
+	if err == nil {
+		rebootDelay = rebootAt.Sub(now)
+	} else {
 		ovr := os.Getenv("SNAPD_REBOOT_DELAY") // for tests
 		if ovr != "" {
 			d, err := time.ParseDuration(ovr)
 			if err == nil {
 				rebootDelay = d
 			}
 		}
-		if err := reboot(rebootDelay); err != nil {
-			return err
-		}
-		// wait for reboot to happen
-		logger.Noticef("Waiting for system reboot")
-		if sigCh != nil {
-			signal.Stop(sigCh)
-			if len(sigCh) > 0 {
-				// a signal arrived in between
-				return nil
-			}
-			close(sigCh)
-		}
-		time.Sleep(rebootWaitTimeout)
-		return fmt.Errorf("expected reboot did not happen")
+		rebootAt = now.Add(rebootDelay)
+		d.state.Set("daemon-system-restart-at", rebootAt)
 	}
-	if d.restartSocket {
-		return ErrRestartSocket
+	return rebootDelay, nil
+}
+
+func (d *Daemon) doReboot(sigCh chan<- os.Signal, waitTimeout time.Duration) error {
+	rebootDelay, err := d.rebootDelay()
+	if err != nil {
+		return err
 	}
+	// ask for shutdown and wait for it to happen.
+	// if we exit snapd will be restared by systemd
+	if err := reboot(rebootDelay); err != nil {
+		return err
+	}
+	// wait for reboot to happen
+	logger.Noticef("Waiting for system reboot")
+	if sigCh != nil {
+		signal.Stop(sigCh)
+		if len(sigCh) > 0 {
+			// a signal arrived in between
+			return nil
+		}
+		close(sigCh)
+	}
+	time.Sleep(waitTimeout)
+	return fmt.Errorf("expected reboot did not happen")
+}
 
+var shutdownMsg = i18n.G("reboot scheduled to update the system")
+
+func rebootImpl(rebootDelay time.Duration) error {
+	if rebootDelay < 0 {
+		rebootDelay = 0
+	}
+	mins := int64(rebootDelay / time.Minute)
+	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), shutdownMsg)
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return osutil.OutputErr(out, err)
+	}
 	return nil
 }
 
+var reboot = rebootImpl
+
 // Dying is a tomb-ish thing
 func (d *Daemon) Dying() <-chan struct{} {
 	return d.tomb.Dying()
 }
 
+func clearReboot(st *state.State) {
+	st.Set("daemon-system-restart-at", nil)
+	st.Set("daemon-system-restart-tentative", nil)
+}
+
+// RebootVerified implements overlord.RestartBehavior.
+func (d *Daemon) RebootVerified(st *state.State, expectedRebootDidNotHappen bool) error {
```

Comment:
Could we use the positive form here? This would avoid the double negative below?